gdal/frmts/netcdf/netcdfsgwriterutil.h

```diff
@@ -68,67 +72,333 @@ namespace nccfdriver
+
+
+        WTransactionLog(WTransactionLog&); // avoid possible undefined behavior
+        WTransactionLog operator=(const WTransactionLog&);
+
+        public:
+            bool logIsNull() { return log == nullptr; }
+            void startLog(); // always call this first to open the file
+            void startRead(); // then call this before reading it
+            void push(std::shared_ptr<OGR_SGFS_Transaction>);
+
+            // read mode
+            std::shared_ptr<OGR_SGFS_Transaction> pop();  // to test for EOF, test to see if pointer returned is null ptr
+
+            // construction, destruction
+            explicit WTransactionLog(std::shared_ptr<std::string> logName);
+            ~WTransactionLog();
+    };
+
+    /* OGR_NCScribe
+     * Buffers several netCDF transactions in memory or in a log.
+     * General scribe class
+     */
+    class OGR_NCScribe
+    {
+        netCDFVID & ncvd;
+        WBuffer buf;
+        WTransactionLog wl;
+        bool singleDatumMode = false;
+
+        std::queue<std::shared_ptr<OGR_SGFS_Transaction>> transactionQueue;
+        std::map<int, size_t> varWriteInds;
+        std::map<int, size_t> varMaxInds;
+        //std::map<int, void*> rawWrite;
+
+        public:
+           /* size_t getWriteCount()
+            * Return the total write count (happened + pending) of certain variable
+            */
+           size_t getWriteCount(int varId) { return this->varMaxInds.at(varId); }
+
+            /* void commit_transaction()
+             * Replays all transactions to disk (according to fs stipulations)
+             */
+           void commit_transaction();
+
+           /* std::shared_ptr<OGR_SGFS_Transaction> pop()
+            * Get the next transaction, if it exists.
+            * If not, it will just return a shared_ptr with nullptr inside
+            */
+           MTPtr pop();
+
+           /* void log_transacion()
+            * Saves the current queued transactions to a log.
+            */
+           void log_transaction();
+
+           /* void enqueue_transaction()
+            * Add a transaction to perform
+            * Once a transaction is enqueued, it will only be dequeued on commit
+            */
+           void enqueue_transaction(std::shared_ptr<OGR_SGFS_Transaction> transactionAdd);
+
+           WBuffer& getMemBuffer() { return buf; }
+
+           /* OGR_SGeometry_Field_Scribe()
+            * Constructs a Field Scribe over a dataset
+            */
+           OGR_NCScribe(netCDFVID& ncd, std::shared_ptr<std::string> name) :
+               ncvd(ncd),
+               wl(name)
+           {}
+
+           /* setSingleDatumMode(...)
+            * Enables or disables single datum mode
+            * DO NOT use this when a commit is taking place, otherwise 
+            * corruption may occur...
+            */
+           void setSingleDatumMode(bool sdm) { this->singleDatumMode = sdm; }
+
+    };
+
+    class ncLayer_SG_Metadata
+    {
+        int & ncID;
```

Comment:
does this need to be a reference ? If so, a comment why wouldn't hurt

Reply:
comment!